File: android/core/src/main/kotlin/com/pocketbot/core/Strategy.kt
package com.pocketbot.core

import kotlin.math.abs
import kotlin.math.min

/**
 * A "EA": mesmo algoritmo de pocket_bot/strategy.py (Python) - reversão do
 * Parabolic SAR como sinal principal, RSI como filtro de exaustão e largura
 * das Bandas de Bollinger como filtro de volatilidade mínima.
 */
object Strategy {

    fun evaluate(pair: String, candles: List<Candle>, cfg: StrategyConfig): Signal? {
        if (candles.size < 3) return null

        val highs = candles.map { it.high }
        val lows = candles.map { it.low }
        val closes = candles.map { it.close }

        val sar = Indicators.psar(highs, lows, cfg.psarAfStep, cfg.psarAfMax)
        val rsiSeries = Indicators.rsi(closes, cfg.rsiPeriod)

        val lastIdx = closes.size - 1
        val lastTrend = closes[lastIdx] > sar[lastIdx]
        val prevTrend = closes[lastIdx - 1] > sar[lastIdx - 1]
        if (lastTrend == prevTrend) return null // só entra logo após a reversão do PSAR

        val lastClose = closes[lastIdx]
        val lastSar = sar[lastIdx]
        val bb = Indicators.bollingerBandsAt(closes, lastIdx, cfg.bbPeriod, cfg.bbStdMult)
        val lastRsi = rsiSeries[lastIdx]

        if (bb.width == null || lastRsi == null || lastClose == 0.0) return null
        if (bb.width < cfg.minBbWidth) return null // volatilidade insuficiente

        val distance = abs(lastClose - lastSar) / lastClose
        // fator de escala empírico para variações típicas de preço nos pares OTC
        val score = min(1.0, distance * 400)

        return if (lastTrend) {
            if (lastRsi >= cfg.rsiUpper) null // alta possivelmente esgotada (sobrecompra)
            else Signal(pair, TradeAction.CALL, score, "PSAR reverteu para alta, sem sobrecompra")
        } else {
            if (lastRsi <= cfg.rsiLower) null // baixa possivelmente esgotada (sobrevenda)
            else Signal(pair, TradeAction.PUT, score, "PSAR reverteu para baixa, sem sobrevenda")
        }
    }

    /** Escolhe, entre os sinais gerados por vários pares no mesmo ciclo, o de maior score. */
    fun pickBest(signals: List<Signal>, cfg: StrategyConfig): Signal? =
        signals.filter { it.score >= cfg.minSignalScore }.maxByOrNull { it.score }
}

File: pocket_bot/strategy.py
"""
A "EA" do bot: reproduz o algoritmo mais comum entre os bots de Pocket
Option pesquisados (ex.: o bot open-source pocket_option_trading_bot, que
usa Parabolic SAR como estratégia principal, e as recomendações recorrentes
de RSI + Bandas de Bollinger como filtro) — entra na direção de uma
reversão recém-detectada do PSAR, descartando o sinal se o RSI já estiver
esgotado no mesmo sentido ou se a volatilidade estiver baixa demais. O
score de confiança é usado para escolher, entre os pares monitorados, qual
operar em cada ciclo.
"""
from dataclasses import dataclass
from typing import Optional

import pandas as pd

from .indicators import bollinger_bands, psar, rsi


@dataclass
class Signal:
    action: str  # "call" ou "put"
    score: float  # 0.0 a 1.0, confiança do sinal
    reason: str
    pair: Optional[str] = None


def evaluate(df: pd.DataFrame, cfg) -> Optional[Signal]:
    high = df["high"].astype(float)
    low = df["low"].astype(float)
    close = df["close"].astype(float)

    if len(close) < 3:
        return None

    sar = psar(high, low, cfg.psar_af_step, cfg.psar_af_max)
    rsi_series = rsi(close, cfg.rsi_period)
    _, _, _, width = bollinger_bands(close, cfg.bb_period, cfg.bb_std_mult)

    trend = (close.reset_index(drop=True) > sar).reset_index(drop=True)
    last_trend = trend.iloc[-1]
    prev_trend = trend.iloc[-2]
    if last_trend == prev_trend:
        return None  # só entra logo após a reversão do PSAR, como o bot de referência

    last_close = close.iloc[-1]
    last_sar = sar.iloc[-1]
    last_width = width.iloc[-1]
    last_rsi = rsi_series.iloc[-1]

    if pd.isna(last_width) or pd.isna(last_rsi) or last_close == 0:
        return None
    if last_width < cfg.min_bb_width:
        return None  # volatilidade insuficiente: mercado "parado", evita operar

    distance = abs(last_close - last_sar) / last_close
    # fator de escala empírico para variações típicas de preço nos pares OTC
    score = min(1.0, distance * 400)

    if last_trend:  # PSAR reverteu para baixo do preço: tendência de alta
        if last_rsi >= cfg.rsi_upper:
            return None  # alta possivelmente esgotada (sobrecompra)
        return Signal("call", score, "PSAR reverteu para alta, sem sobrecompra")

    if last_rsi <= cfg.rsi_lower:
        return None  # baixa possivelmente esgotada (sobrevenda)
    return Signal("put", score, "PSAR reverteu para baixa, sem sobrevenda")
